File: leetcode/0565_array_nesting.cpp
// Copyright (c) 2018 Jacopo Notarstefano

#include <algorithm>
#include <cassert>
#include <vector>

using namespace std;

class Solution {
 public:
    int arrayNesting(vector<int>& nums) {
        int result = 0;

        for (auto num : nums) {
            if (num == -1) {
                continue;
            }
            int current = num, length = 0;
            while (nums[current] != -1) {
                int previous = current;
                current = nums[current];
                nums[previous] = -1;
                length++;
            }
            result = max(result, length);
        }

        return result;
    }
};

int main() {
    auto solution = Solution();

    vector<int> nums = {5, 4, 0, 3, 1, 6, 2};

    assert(4 == solution.arrayNesting(nums));
}

File: leetcode/0565_array_nesting.py
# -*- coding: utf-8 -*-


class Solution:
    def arrayNesting(self, nums):
        result = 0

        for num in nums:
            if num is None:
                continue
            current, length = num, 0
            while nums[current] is not None:
                previous = current
                current = nums[current]
                nums[previous] = None
                length += 1
            result = max(result, length)

        return result


if __name__ == "__main__":
    solution = Solution()

    assert 4 == solution.arrayNesting([5, 4, 0, 3, 1, 6, 2])
